DSA-Campaign-Uplift-Estimation/src/main/java/com/google/sps/servlets/BlackBoxServlet.java

```diff
@@ -0,0 +1,125 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.classes.KeywordCampaign;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import java.io.IOException;
+import java.io.PrintWriter;
+import com.google.gson.Gson;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+
+// runs all the pending DSA campaigns through the black box and updates estimation results
+@WebServlet("/black-box")
+public class BlackBoxServlet extends HttpServlet {
+    
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {	
+        // get all the pending DSA campaigns from datastore
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("DSACampaign").setFilter(new Query.FilterPredicate("campaignStatus", Query.FilterOperator.EQUAL, "pending"));
+        PreparedQuery results = datastore.prepare(query);
+        
+        for (Entity DSACampaignEntity : results.asIterable()) {
+            // run the DSA campaign through the black box
+            blackbox(DSACampaignEntity);
+        }
+
+        response.sendRedirect("/Home/home.html");
+    }
+
+    /*
+     * Runs the DSA campaign and obtains estimation results.
+     * Updates the DSA campaign entity with the estimation results and changes the campaign status from pending to complete.
+     * Implementation explanations are in the design doc.
+     */
+    public static void blackbox(Entity DSACampaignEntity) {
+        // retrieve the DSA campaign's corresponding keyword campaign from datastore
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("keywordCampaign").setFilter(new Query.FilterPredicate("keywordCampaignId", Query.FilterOperator.EQUAL, (String) DSACampaignEntity.getProperty("keywordCampaignId")));
+    	Entity keywordCampaignEntity = datastore.prepare(query).asSingleEntity();
```

Comment:
.asSingleEntity() will cause a crash, so we need some type of error handling for duplicate keywordCampaignId's.

Reply:
Added a function to create unique keyword campaign id's.